### ISSUE TITLE
Dropdowns should not respond to mouse wheel movement until they are open

### ISSUE THREAD

**Issue by spilth** [qgis/QGIS#38572]
**Describe the bug**

3 times today I have changed my Point layer's Symbol type from Categorized to something else when accidentally hitting my mouse's scroll wheel while passing over the Symbol type dropdown. This results in me losing all the customization I have made to my Categorized Symbols - super frustrating.

**How to Reproduce**

1. Create a map with a Point layer
2. Set the Point layer's Symbology to Categorized
3. Spend several minutes customizing a bunch of different symbol types based on different values
4. Hover over the Symbology dropdown and scroll your mouse wheel
5. Change it back to Categorized and notice that all your work is gone
6. Cry

**QGIS and OS versions**

```
QGIS version | 3.14.15-Pi | QGIS code revision | d5114d2cfa
-- | -- | -- | --
Compiled against Qt | 5.12.3 | Running against Qt | 5.12.3
Compiled against GDAL/OGR | 2.4.1 | Running against GDAL/OGR | 2.4.1
Compiled against GEOS | 3.7.2-CAPI-1.11.2 | Running against GEOS | 3.7.2-CAPI-1.11.2 b55d2125
Compiled against SQLite | 3.28.0 | Running against SQLite | 3.28.0
PostgreSQL Client Version | 11.3 | SpatiaLite Version | 4.3.0a
QWT Version | 6.1.4 | QScintilla2 Version | 2.11.1
Compiled against PROJ | 5.2.0 | Running against PROJ | Rel. 5.2.0, September 15th, 2018
OS Version | macOS 10.15
Active python plugins | batch_gps_importer; processing; db_manager; MetaSearch
```

**Additional context**

I am using a Magic Mouse on macOS where it's pretty easy to accidentally scroll the mouse wheel.

That said, no other application treats dropdown this way. The mouse wheel doesn't do anything even when you click the Dropdown open. Perhaps this is just how QT works but it's pretty frustrating.

I suppose I could disable my mouse wheel completely, but considering no other application handles dropdowns this way....

**Comment by spilth** [comment]
I am not familiar with QT, but these seems like it might allow you to control this behavior: https://doc.qt.io/qt-5/qml-qtquick-controls2-control.html#focusPolicy-prop

**Comment by andreasneumann** [comment]
I agree that this is a "general" problem throughout QGIS (in all places where there are comboboxes). It also happened to me while editing data, that I accidentally changed data in an attribute form, because of this behavior. I agree with @spilth that it would be better to only activate mouse wheel signals if the combobox is unfolded.

@m-kuhn or @3nids - do you have any idea?

**Comment by m-kuhn** [comment]
I agree it's desirable and also doable (see https://stackoverflow.com/a/43467364/2319028)

**Comment by nyalldawson** [comment]
@m-kuhn there's an easier approach -- you can mess with ScrollAreaFilter::eventFilter from qgsscrollarea.cpp and add an event filter to hide the wheel events from combobox targets there.

**Comment by m-kuhn** [comment]
this which will make combobox behave differently when in scrollboxes and when not, no?

**Comment by uclaros** [comment]
This behavior affects other widgets too, like spinboxes.
See relevant #34762

**Comment by nyalldawson** [comment]
@m-kuhn 

> this which will make combobox behave differently when in scrollboxes and when not, no?

True -- but I don't think switching everything to a custom subclass is an option either. Maybe we could install a recursive event filter on the application level which also responds to child added events to install itself on all QComboBox widgets (but not sure if that is too costly...)

**Comment by nyalldawson** [comment]
@uclaros 
> This behavior affects other widgets too, like spinboxes.

You can prise the scrollwheel-change-spinbox-value feature from my cold dead hands. I use that 100/1000s of times a day, and is a FANTASTIC feature.

I agree the combo box changing via scroll wheel is silly, but for some widgets (spin boxes, sliders) it's definitely a feature.

**Comment by uclaros** [comment]
> You can prise the scrollwheel-change-spinbox-value feature from my cold dead hands. I use that 100/1000s of times a day, and is a FANTASTIC feature.

I was hoping for a reply like that!! :) It is indeed a fantastic feature and I too use it all the time!

But, you know what else is a FANTASTIC feature that I also use daily? You guessed right, changing combo box values with the scroll wheel, without having to click-to-open and click-to-select. It is the same as a spinbox, though instead of sequential numbers it has strings! When you already know the contents of the combo box, it's EXACTLY the same!

Changing widget values with the scroll wheel has two drawbacks:
1. Unintentional value change while trying to scroll the long dialog
- This was partially fixed by locking the widgets while scrolling and unlocking with a delay afterwards. However, if the cursor is over a widget when you try to initiate the scrolling, its value is messed!
2. It is common that a combo box value change triggers more changes that can't be undone easily.
- Such is the case that @spilth mentions on the report. The real problem here is not that the user unintentionally changed the combo box value. The real problem is that he lost his categories, and that could have happened by changing the combo box value by clicking too (has happened to me too more than once).

So, what I see as a potential solution is to implement one or both of:
1. Require pressing a modifier key in order to change values with the scroll wheel on ANY type of widget. (I don't really like that but would get used to it. Reportedly Blender works that way so it's a tested solution)
2. Don't reset the styling page when changing the symbol type, re-applying previous settings when the combo box changes back. (I really like that, it's quite common to have a categorized symbology but temporarily I want to see all categories with the same styling and it's faster than creating a new style).

**Comment by nyalldawson** [comment]
> Require pressing a modifier key in order to change values with the scroll wheel on ANY type of widget. (I don't really like that but would get used to it. Reportedly Blender works that way so it's a tested solution)

I could live with that too. It'd have to be a non-ctrl key shortcut though, that's already used for "fine" increments to the value (to match canvas zoom behavior)

**Comment by spilth** [comment]
> So, what I see as a potential solution is to implement one or both of:
> 
> 1. Require pressing a modifier key in order to change values with the scroll wheel on ANY type of widget. (I don't really like that but would get used to it. Reportedly Blender works that way so it's a tested solution)
> 2. Don't reset the styling page when changing the symbol type, re-applying previous settings when the combo box changes back. (I really like that, it's quite common to have a categorized symbology but temporarily I want to see all categories with the same styling and it's faster than creating a new style).

I prefer 1 over 2 because it prevents me from accidentally changing a LOT of input elements. I believe there are even some Font selector buttons that change the size of the font if you scroll your mouse wheel over it. 🤦‍♂️ 

The place I keep running into these sorts of accidental changes is in the Layer Styling panel when trying to use the mouse wheel to get to UI elements further down the panel. But I would welcome this kind of control most everywhere.

**Comment by nyalldawson** [comment]
I'm thinking we should just make this a global option and do away with the modifier requirement.

**Comment by m-kuhn** [comment]
@nyalldawson what would the option be, "disable mouse wheel on range widgets" or "disable mouse wheel on range widgets and combo boxes" or something else?

**Comment by nyalldawson** [comment]
@m-kuhn 

How about "allow widget values to be changed via mouse wheel actions" ?

**Comment by m-kuhn** [comment]
@nyalldawson so you are happy if this is an "all or nothing" option, I understood you would like to [disable this for combo boxes but not for spin boxes](https://github.com/qgis/QGIS/issues/38572#issuecomment-687934787)?

**Comment by uclaros** [comment]
Why not "Require modifier key to change widget values with the mouse wheel." Best of both worlds.

**Comment by spilth** [comment]
Any updates on this issue? It just had it happened to me once again (losing all my Layer Style changes by accidentally scrolling over the combo box). Fortunately I had a recent previous save to revert back, but that's a really rough way to work.

**Comment by M-Rick** [comment]
The Magic Mouse has another issue, it is in contrary lacking of reactivity/efficiency to scroll in dialog windows, such as the properties or the the processing tools.

**Comment by AlisterH** [comment]
Perhaps an alternative solution could be a "safe scrolling" feature where you can hold a modifier key to safely scroll a dialog, and it temporarily "locks" widgets like comboboxes spinboxes and sliders.
The major downside is that it would only help people that actually know about the modifier key.  But I reckon it would seem quite natural when you do know about it.

**Comment by spilth** [comment]
> The Magic Mouse has another issue, it is in contrary lacking of reactivity/efficiency to scroll in dialog windows, such as the properties or the the processing tools.

That should probably be opened up as a separate issue.

**Comment by kannes** [comment]
I would love to have this with a modifier key. Being able to use the scroll wheel for fast tweaking is insanely useful and fun but accidentally changing values when you just wanted to scroll a panel is insanely annoying and painful. So globally disabling the option would take away lots of usability.

An easy modifier key (if a good one is still available, Shift? Ctrl+Shift?) would be fantastic, I am sure it would be quick to internalize to muscle memory.

**Comment by Hedaja** [comment]
I have been stumbeling across that issue using QGIS quite frequently by now. I enjoy using the scroll wheel for counting up numbers but I also defiently have accidentally changed numbers by trying to scroll through an input form. Luckily I usually quickly notice but I might still have to discard my edit if I don't remember the old number anymore.
As people already mentioned I think using a modifier key for changing values/dropdown lists would be a good solution.

> An easy modifier key (if a good one is still available, Shift? Ctrl+Shift?) would be fantastic, I am sure it would be quick to internalize to muscle memory.

I gave it a try on QGIS 3.40.6. crtl key as people have mentioned sometimes seems to be used for "fine increments". Not quite sure what the alt key does. In integer/text fields it seems to select the whole value but it also disables the scroll to change function (so kinda what would be useful but in reverse).
Shift key seems to be unused right now and doesn't seem to modify the scrolling behavior in any way. So that might be a good option.

As someone mentioned Blender above, it use ctrl key to adjust values and ctrl+shift for fine increments.

**Comment by kolamalursurvey** [comment]
I was hoping to see this bug resolved in QGIS 4.0, turned out only to be an empty hope. 

I work at a large company with very complex spatial datasets, and this specific issue makes me hesitant to introduce QGIS to a newbie users. I can't imagine if someone accidentally changing very-critical values without realizing it, only because they use scroll to navigate a menu without clicking anything.

**Comment by m-kuhn** [comment]
@kolamalursurvey apparently this is an issue specifically with QGIS 4 (see https://github.com/qgis/QGIS/pull/65656#issuecomment-4177239995 where @uclaros describes this in more detail).

Can you check if this problem is also present for you for QGIS 3.44 (which is about to become LTR)?

**Comment by Hedaja** [comment]
> [@kolamalursurvey](https://github.com/kolamalursurvey) apparently this is an issue specifically with QGIS 4 (see [#65656 (comment)](https://github.com/qgis/QGIS/pull/65656#issuecomment-4177239995) where [@uclaros](https://github.com/uclaros) describes this in more detail).
> 
> Can you check if this problem is also present for you for QGIS 3.44 (which is about to become LTR)?

The comment you mention is about a bug in the linked pull request that leads to scrolling stalling on items.

But the issue is that in QGIS you can accidentally change values when just scrolling through a list. Unfortunately this seems to have been a long time issue. I have it in QGIS 3.40 and I also just tested again in 3.44.9. with the same result
Here another quick video where I demonstrate trying to scroll up and down a list and accidentally changing the values instead.

EDIT: While playing around I just noticed that there seems to be a behavior implemented that is supposed to prevent the accidental change. When scrolling next to the list and then hovering the fields, it seem to not change the value until you pause and move the mouse over one of the fields. If you pause and not move the mouse before scrolling, it stays in the list scrolling mode.
I see the idea behind it but I still feel like this is very error prone because it's really hard to influence the behavior (one small mouse movement could mean accidentally changing a value and scrolling through the list. I think having a dedicated modifier key (as for example in Blender) would be useful.

https://github.com/user-attachments/assets/09feaec9-5f24-4dff-810f-297339283c74

<table width='100%' align='center'><tr><td>QGIS-Version</td><td>3.44.9-Solothurn</td></tr><tr><td>QGIS-Codeversion</td><td><a href="https://github.com/qgis/QGIS/commit/0bf732d52fb">0bf732d52fb</a></td></tr><tr><td colspan="2">&nbsp;</td></tr><tr><td colspan="2"><b>Bibliotheken</b></td></tr><tr><td>Qt-Version</td><td>5.15.13</td></tr><tr><td>Python-Version</td><td>3.12.3</td></tr><tr><td>GDAL-Version</td><td>3.8.4</td></tr><tr><td>PROJ-Version</td><td>9.4.0</td></tr><tr><td>EPSG-Registraturdatenbankversion</td><td>v11.004 (2024-02-24)</td></tr><tr><td>GEOS-Version</td><td>3.12.1-CAPI-1.18.1</td></tr><tr><td>SQLite-Version</td><td>3.45.1</td></tr><tr><td>PostgreSQL-Client-Version</td><td>16.13</td></tr><tr><td>SpatiaLite-Version</td><td>5.1.0</td></td></tr><tr><td>QWT-Version</td><td>6.1.4</td></tr><tr><td>QScintilla2-Version</td><td>2.14.1</td></tr><tr><td>BS-Version</td><td>Linux Mint 22.3</td></tr><tr><td colspan="2">&nbsp;</td></tr><tr><td colspan="2"><b>Aktive Python-Erweiterungen</b></td></tr><tr><td>QuickOSM</td><td>2.5.1</td></tr><tr><td>qfieldsync</td><td>v4.19.0</td></tr><tr><td>quick_map_services</td><td>1.3.0</td></tr><tr><td>DigitizingTools</td><td>1.5.3</td></tr><tr><td>NLWKN Brutvogelkartierung</td><td>2.7.5</td></tr><tr><td>SRTM-Downloader</td><td>3.3.1</td></tr><tr><td>batchvectorlayersaver</td><td>0.9</td></tr><tr><td>brutvoegel_papierrevier_lr</td><td>1.1</td></tr><tr><td>AutoLayoutTool</td><td>2.2</td></tr><tr><td>processing_saga_nextgen</td><td>1.1.0</td></tr><tr><td>qgis-maptiler-plugin</td><td>3.6</td></tr><tr><td>AppendFeaturesToLayer</td><td>2.2.0</td></tr><tr><td>clipper</td><td>1.2</td></tr><tr><td>ProjectPackager</td><td>0.7.6</td></tr><tr><td>db_manager</td><td>0.1.20</td></tr><tr><td>MetaSearch</td><td>0.3.6</td></tr><tr><td>grassprovider</td><td>2.12.99</td></tr><tr><td>processing</td><td>2.12.99</td></tr></table>

**Comment by m-kuhn** [comment]
> The comment you mention is about a bug in the linked pull request that leads to scrolling stalling on items.

It's about a bug in QGIS 4.0 (which is also visible in this pull request).

> EDIT: While playing around I just noticed that there seems to be a behavior implemented that is supposed to prevent the accidental change. When scrolling next to the list and then hovering the fields, it seem to not change the value until you pause and move the mouse over one of the fields. If you pause and not move the mouse before scrolling, it stays in the list scrolling mode. I see the idea behind it but I still feel like this is very error prone because it's really hard to influence the behavior (one small mouse movement could mean accidentally changing a value and scrolling through the list. I think having a dedicated modifier key (as for example in Blender) would be useful.

@nyalldawson and @uclaros are you okay with the suggested approach (modifier key) which has also seen several upvotes here https://github.com/qgis/QGIS/issues/38572#issuecomment-1307303037 ?

**Comment by uclaros** [comment]
I stand by my original opinion on this thread.
1. Global option to _Require modifier key to change widget values with the mouse wheel._
2. Apply to __all widgets that change values on wheel events__, not only comboboxes. Spinboxes, dials, sliders, color widgets, symbol buttons, potentially other widgets I am missing.

**Comment by m-kuhn** [comment]
So the modifier key makes sense, applying this to all affected widgets obviously as well.

What I am wondering is if we need another global option for this, which would mean to have one more option on user interface level and to keep (and fix) the current timer logic. Are there one or two other people who would also prefer to keep this behavior available?

**Comment by nyalldawson** [comment]
I'm ok with a modifier approach and skip the setting. The modifier CANNOT be ctrl though, that already has meaning when scrolling on a spin box.

I'm interested to hear how'd you go about this. I'd rather not have to replace all our QComboBox instances everywhere with a custom class 😱  (not to mention that plugins will never do that, so we'd still have the original issue sitting around). Do you think it's achievable via some event filter magic at an application level?

**Comment by signedav** [comment]
> I'm ok with a modifier approach and skip the setting. The modifier CANNOT be ctrl though, that already has meaning when scrolling on a spin box.

Because I was not aware how Ctrl works in the spinbox , I leave the note here: It turns off the steps. For example (when having a step of 10 defined), scrolling without Ctrl changes the value by 10 (0, 10, 20 ...). Pressing Ctrl changes it by 1 instead (21, 22, 23 ...).

**Comment by nyalldawson** [comment]
@signedav just to clarify -- it doesn't turn off the steps, it just reduces them to 1/10th of the usual step

**Comment by signedav** [comment]
> [@signedav](https://github.com/signedav) just to clarify -- it doesn't turn off the steps, it just reduces them to 1/10th of the usual step

Ah, thanks for that info. Unfortunately I tested exactly with a step of 10 🤦 

- Means 1/10 and rounded down (means 17 as step would make step of 1) on scrolling
- on pressing up and down with the keyboard arrows it multiplies by 10 (means on 17 it's 170, 340 etc)
